### ISSUE TITLE
Override TOTAL in configuration

### ISSUE THREAD

**Issue by tYYGH** [rclone/rclone#3997]
#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.51.0
- os/arch: linux/amd64
- go version: go1.13.7
```

#### What problem are you are trying to solve?

Cloud providers bill based on usage, and crossing a usage threshold can provoke an automatic switch to the upper (and costlier) offer.

#### How do you think rclone should be changed to solve that?

I’ve searched the issues for my problem, and my issue is exactly #1956. It is closed, though.
I totally agree with the NO-FIX resolution, since the cloud providers just don’t provide the necessary data.

However, apparently Rclone now has the feature, at least for those providers that do provide the information: I did some tests, and both FREE and TOTAL spaces are correctly reported for:
* Google Drive
* Koofr
* MS OneDrive

Strangely, even though both FREE and TOTAL are correctly reported with the `about` command, I get wrong results with the `mount` command for:
* Mega: I get a TOTAL of 148G instead of 50G, and 13G FREE insead of 4G;
* Yandex: while the TOTAL is correct (10G), the FREE space is reported as 1P.

So, what I suggest is adding the possibility to configure an optional per-remote TOTAL that would be used for that remote in place of whatever is reported by the remote.
If the TOTAL is such configured for a remote, then the FREE space would be the simple calculation `TOTAL−USED=FREE`.

**Comment by tYYGH** [comment]
Although I’m just discovering rclone, and I don’t know the Go language, I’m trying to see if I’d be able to bring the change myself.

I think I’m trying to get some bearings, but I’d be glad to get some help.
If I understand correctly:
* Based on `fs.go` found in `cmd/cmount/` and `cmd/mount/`, I think that the computation I’m interested in happens in `VFS.Statfs`, which itself (`vfs/vfs.go`) is using the `About` “feature” of the underlying backend.
* In `mega.go` for example, `features` is mostly filled by `Fill` in `fs/fs.go`, which sets `About` to the `About` function found in `mega.go` as an implementation of the `Abouter` interface.
* `fs/config/config.go` is the handling of the configuration. I appended that to the `const {…}` block:
```go
	// ConfigForceTotalFsSize is an optional override of the Total size reported by a filesystem
	ConfigForceTotalFsSize = "fs_total_bytes"
```
Then I should just have to call `FileGet(_a section_, ConfigForceTotalFsSize)` at the right places (i.e. for each backend)…

**What I still don’t understand, is why the data reported by the Linux command `df` does not match that reported by the rclone command `about` for some filesystems (eg. Mega).**

I tried the following patch:
```diff
diff --git a/backend/mega/mega.go b/backend/mega/mega.go
index 1802fa962..b4d4ba832 100644
--- a/backend/mega/mega.go
+++ b/backend/mega/mega.go
@@ -89,6 +89,11 @@ permanently delete objects instead.`,
                        // Encode invalid UTF-8 bytes as json doesn't handle them properly.
                        Default: (encoder.Base |
                                encoder.EncodeInvalidUtf8),
+               }, {
+                       Name:     config.ConfigForceTotalFsSize,
+                       Help:     config.ConfigForceTotalFsSizeHelp,
+                       Advanced: true,
+                       Default:  0,
                }},
        })
 }
@@ -100,6 +105,7 @@ type Options struct {
        Debug      bool                 `config:"debug"`
        HardDelete bool                 `config:"hard_delete"`
        Enc        encoder.MultiEncoder `config:"encoding"`
+       ForcedMax  uint64               `config:"fs_total_bytes"`
 }
 
 // Fs represents a remote mega
@@ -912,12 +918,22 @@ func (f *Fs) About(ctx context.Context) (*fs.Usage, error) {
        if err != nil {
                return nil, errors.Wrap(err, "failed to get Mega Quota")
        }
-       usage := &fs.Usage{
-               Total: fs.NewUsageValue(int64(q.Mstrg)),           // quota of bytes that can be used
-               Used:  fs.NewUsageValue(int64(q.Cstrg)),           // bytes in use
-               Free:  fs.NewUsageValue(int64(q.Mstrg - q.Cstrg)), // bytes which can be uploaded before reaching the quota
+       if f.opt.ForcedMax > 0 {
+               var used = q.Cstrg                                     // bytes in use
+               usage := &fs.Usage{
+                       Total: fs.NewUsageValue(int64(f.opt.ForcedMax)),   // quota of bytes that can be used
+                       Used:  fs.NewUsageValue(int64(used)),
+                       Free:  fs.NewUsageValue(int64(f.opt.ForcedMax - used)), // bytes which can be uploaded before reaching the quota
+               }
+               return usage, nil
+       } else {
+               usage := &fs.Usage{
+                       Total: fs.NewUsageValue(int64(q.Mstrg)),           // quota of bytes that can be used
+                       Used:  fs.NewUsageValue(int64(q.Cstrg)),           // bytes in use
+                       Free:  fs.NewUsageValue(int64(q.Mstrg - q.Cstrg)), // bytes which can be uploaded before reaching the quota
+               }
+               return usage, nil
        }
-       return usage, nil
 }
 
 // ------------------------------------------------------------
diff --git a/fs/config/config.go b/fs/config/config.go
index c7793f412..d5ac0cf9a 100644
--- a/fs/config/config.go
+++ b/fs/config/config.go
@@ -72,6 +72,12 @@ const (
 
        // ConfigAuthNoBrowser indicates that we do not want to open browser
        ConfigAuthNoBrowser = "config_auth_no_browser"
+
+       // ConfigForceTotalFsSize is an optional override of the Total size reported by a filesystem
+       ConfigForceTotalFsSize = "fs_total_bytes"
+
+       // ConfigForceTotalFsSizeHelp is the help for ConfigForceTotalFsSize
+       ConfigForceTotalFsSizeHelp = "Force a maximum capacity in bytes, regardless of the capacity reported by the backend."
 )
 
 // Global
```

After that, I reconfigured my Mega backend, with a Max of 49G instead of 50G, and the `about` command correctly reported 1G less. However, the values reported by `df` are still wrong…

**Comment by tYYGH** [comment]
Oops :laughing: 
Actually what I changed (for Mega) works, and `df` agrees with `about`!
Mega is so _slow_, that I was just too quick in running `df`, which actually measured my `/home` partition!

So, new state of things:
* Linux `df` and rclone `about` agree with each other on: Google Drive, Koofr, Mega, MS OneDrive.
* As for Yandex, it still reports 1PB of Free space even though the Total space is 10G… which agrees with `about`, for which the Free amount is not given.
* Last but not least, `df` and `about` also agree on Hubic, for which neither Total nor Free is given, and thus appears to have 1PB of Total and Free space.

Next step: I’ll try and apply my patch to Yandex and Hubic, and see if it gives the appropriate result. _And_, most of all, I’ll have to see if it has any impact on the success of writing a new file above the chosen limit!

**Comment by ncw** [comment]
> * Linux `df` and rclone `about` agree with each other on: Google Drive, Koofr, Mega, MS OneDrive.

Great!

> * As for Yandex, it still reports 1PB of Free space even though the Total space is 10G… which agrees with `about`, for which the Free amount is not given.

That 1PB figure is "unlimited" and is what rclone uses in `df` when the total space isn't known.

I had a quick look at the backend code and it looks like it *should* work! Test with `rclone about yandex: -vv --dump bodies` to see what is actually returned

> * Last but not least, `df` and `about` also agree on Hubic, for which neither Total nor Free is given, and thus appears to have 1PB of Total and Free space.
> 
> Next step: I’ll try and apply my patch to Yandex and Hubic, and see if it gives the appropriate result. _And_, most of all, I’ll have to see if it has any impact on the success of writing a new file above the chosen limit!

I think your approach of patching the individual backends is the wrong one.

You should be patching the VFS layer, so add a new command line flag here - `vfs-total-size` say: https://github.com/rclone/rclone/blob/master/vfs/vfsflags/vfsflags.go

And do the work in the Statfs function

https://github.com/rclone/rclone/blob/ac9cb50fdb9aa06342d58da61dd15687ade2e8d8/vfs/vfs.go#L481

That will then work for all users of the VFS layer without patching any of the backends. It won't work for `rclone about` but I don't think that is important here.

**Comment by tYYGH** [comment]
> That 1PB figure is "unlimited" and is what rclone uses in `df` when the total space isn't known.

Yes, I figured as much, but it’s always nice to have confirmation :-)

> I had a quick look at the backend code and it looks like it _should_ work! Test with `rclone about yandex: -vv --dump bodies` to see what is actually returned

Looking at this dump reminded me of a “detail” I had forgotten, but which has its importance here: I’m not actually using the `yandex` backend, but instead the `webdav` backend, because apparently Yandex blacklisted the rclone token for “bad behaviour”. Thus I cannot say what the `yandex` backend would do. For what it’s worth, here’s the dump for `webdav`:

```
2020/02/25 18:25:55 DEBUG : rclone: Version "v1.51.0" starting with parameters ["rclone" "about" "yandex-yves:" "-vv" "--dump" "bodies"]
2020/02/25 18:25:55 DEBUG : Using config file from "/home/yves/.config/rclone/rclone.conf"
2020/02/25 18:25:55 DEBUG : You have specified to dump information. Please be noted that the Accept-Encoding as shown may not be correct in the request and the response may not show Content-Encoding if the go standard libraries auto gzip encoding was in effect. In this case the body of the request will be gunzipped before showing it.
2020/02/25 18:25:55 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2020/02/25 18:25:55 DEBUG : HTTP REQUEST (req 0xc00024a900)
2020/02/25 18:25:55 DEBUG : PROPFIND / HTTP/1.1
Host: webdav.yandex.com
User-Agent: rclone/v1.51.0
Content-Length: 139
Authorization: XXXX
Depth: 0
Referer: https://webdav.yandex.com/
Accept-Encoding: gzip

<?xml version="1.0" ?>
<D:propfind xmlns:D="DAV:">
 <D:prop>
  <D:quota-available-bytes/>
  <D:quota-used-bytes/>
 </D:prop>
</D:propfind>
2020/02/25 18:25:55 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2020/02/25 18:25:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2020/02/25 18:25:56 DEBUG : HTTP RESPONSE (req 0xc00024a900)
2020/02/25 18:25:56 DEBUG : HTTP/1.1 207 Multi-Status
Content-Length: 316
Content-Type: application/xml; charset=UTF-8
Date: Tue, 25 Feb 2020 17:25:56 GMT
Server: Jetty(9.4.11.v20180605)
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
Yandex-Cloud-Request-Id: dav-45f61541f3c063f3bb29d4620726e58b-myt6-ef8336f0a848
Yandex-Uid: 347488865

<?xml version="1.0" encoding="UTF-8"?><d:multistatus xmlns:d="DAV:"><d:response><d:href>/</d:href><d:propstat><d:status>HTTP/1.1 200 OK</d:status><d:prop><d:quota-used-bytes>94496452</d:quota-used-bytes><d:quota-available-bytes>10642921788</d:quota-available-bytes></d:prop></d:propstat></d:response></d:multistatus>
2020/02/25 18:25:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Total:   10G
Used:    90.119M
2020/02/25 18:25:56 DEBUG : 5 go routines active
2020/02/25 18:25:56 DEBUG : rclone: Version "v1.51.0" finishing with parameters ["rclone" "about" "yandex-yves:" "-vv" "--dump" "bodies"]
```

IMHO, the `webdav` backend _should_ also give information about Free space, considering the available information…

> > Next step: I’ll try and apply my patch to Yandex and Hubic, and see if it gives the appropriate result. _And_, most of all, I’ll have to see if it has any impact on the success of writing a new file above the chosen limit!
> 
> I think your approach of patching the individual backends is the wrong one.
> 
> You should be patching the VFS layer, so add a new command line flag here - `vfs-total-size` say: https://github.com/rclone/rclone/blob/master/vfs/vfsflags/vfsflags.go
> 
> And do the work in the Statfs function
> 
> https://github.com/rclone/rclone/blob/ac9cb50fdb9aa06342d58da61dd15687ade2e8d8/vfs/vfs.go#L481
> 
> That will then work for all users of the VFS layer without patching any of the backends. It won't work for `rclone about` but I don't think that is important here.

I totally agree, both with the method, and with that last remark.
However, the reason I started patching at the backend level, is because I wish to have an individual override for each backend, and I thought, the only place where I would have access to this setting (`f.opt.ForcedMax` in the above patch) is in the backend itself. Do you think I have a way to have that in the VFS layer? (pseudo-code:)
```
if (*actual_backend).config.ForcedMax {
  …
```
Also, what would guarantee to me, that the name of the parameter would be the same for each backend?

**Comment by tYYGH** [comment]
I did exactly as described above, and it works with `mount` (but not with `about` as you foretold):
```diff
-hubic-yves:      fuse.rclone   1,0P    880M  1,0P   1% /home/yves/Nextcloud/dev/rclone/test-mount
+hubic-yves:      fuse.rclone    25G    880M   25G   4% /home/yves/Nextcloud/dev/rclone/test-mount
```
However, this is informative only, and does not actually limit anything :-(
I tried with a smaller limit, and could copy a too-big file:

```bash
$ ~/go/bin/rclone mount hubic-yves: test-mount/ &
$ df -hT test-mount/
hubic-yves:      fuse.rclone  1000M    880M  121M  88% /home/yves/Nextcloud/dev/rclone/test-mount
$ ls -l ~/GeoGebra-Linux-Portable-5-0-451-0.tar.bz2
-rw-r--r-- 1 yves users 136483130  1 avril  2018 /home/yves/GeoGebra-Linux-Portable-5-0-451-0.tar.bz2
$ cp ~/GeoGebra-Linux-Portable-5-0-451-0.tar.bz2 test-mount/default/
$ df -hT test-mount/
hubic-yves:      fuse.rclone  1000M    -64Z  1,0P 101% /home/yves/Nextcloud/dev/rclone/test-mount
```
Any idea on the best way/place to enforce the configured maximum?

**Comment by ncw** [comment]
> IMHO, the `webdav` backend _should_ also give information about Free space, considering the available information

That looks like a bug!

Try this patch

```go
diff --git a/backend/webdav/webdav.go b/backend/webdav/webdav.go
index 96bf488c7..6cba7821f 100644
--- a/backend/webdav/webdav.go
+++ b/backend/webdav/webdav.go
@@ -992,6 +992,7 @@ func (f *Fs) About(ctx context.Context) (*fs.Usage, error) {
 	if q.Available != 0 || q.Used != 0 {
 		if q.Available >= 0 && q.Used >= 0 {
 			usage.Total = fs.NewUsageValue(q.Available + q.Used)
+			usage.Free = fs.NewUsageValue(q.Available)
 		}
 		if q.Used >= 0 {
 			usage.Used = fs.NewUsageValue(q.Used)
```

> However, the reason I started patching at the backend level, is because I wish to have an individual override for each backend, and I thought, the only place where I would have access to this setting (f.opt.ForcedMax in the above patch) is in the backend itself.

At some point I'd like to be able to override config in the backend configuration which would solve your problem.

However since rclone can only run one mount at once, then I don't see a problem with supplying a `--vfs-total-size` flag for each one?

> However, this is informative only, and does not actually limit anything :-(

Yes indeed! The VFS layer doesn't check the size before writing, it relies on the backend to do this.

Why do you need it to actually limit the writes? I thought this was only a cosmetic feature?

**Comment by tYYGH** [comment]
> Try this patch
> […]

I did this instead, before you proposed an alternative:
```diff
--- a/vfs/vfs.go
+++ b/vfs/vfs.go
@@ -506,4 +506,13 @@ func (vfs *VFS) Statfs() (total, used, free int64) {
                if u.Used != nil {
                        used = *u.Used
                }
+               if total < 0 && free >= 0 && used >= 0 {
+                       total = free + used
+               }
+               if free < 0 && total >= 0 && used >= 0 {
+                       free = total - used
+               }
+               if used < 0 && total >= 0 && free >= 0 {
+                       used = total - free
+               }
        }
```
IMHO, it seems more generic, is it not?
> 
> > However, the reason I started patching at the backend level, is because I wish to have an individual override for each backend, and I thought, the only place where I would have access to this setting (f.opt.ForcedMax in the above patch) is in the backend itself.
> 
> At some point I'd like to be able to override config in the backend configuration which would solve your problem.
> 
> However since rclone can only run one mount at once, then I don't see a problem with supplying a `--vfs-total-size` flag for each one?

Good point! I’ll follow your lead on this one :-)

At this point, I probably should fork the project and prepare proper pull-requests, starting with the above patch.

> > However, this is informative only, and does not actually limit anything :-(
> 
> Yes indeed! The VFS layer doesn't check the size before writing, it relies on the backend to do this.
> 
> Why do you need it to actually limit the writes? I thought this was only a cosmetic feature?

Hardly so! I quote myself here (initial message): “Cloud providers bill based on usage, and crossing a usage threshold can provoke an automatic switch to the upper (and costlier) offer.”

I wish to avoid storing more data than I can afford on a single remote, taking into account the fact that I’ll balance my data among several remotes (probably using mhddfs, or mergerfs).

**Comment by tYYGH** [comment]
Hello, here's a small recap’ of the whole subject:

* Each mounted filesystem must report all of Free, Used, and Total space. With the pull-requests I wrote, this is taken care of (well, one PR is still on my side: `--vfs-forced-total=XXX`).
* Each operation on a mounted filesystem must heed the computed Free and Total space. _That_ is what remains to be done.

Note: I’ll develop this patch, with the pre-requisite (that will be documented), that the mount-point is the only access to the backend while it is mounted; i.e. _all_ operations happen locally. Maybe someday someone with better knowledge of the code-base and the Go language will be able to remove this limitation…

So, basically, what I want to achieve, is that each Copy/Delete operation does this:

1. read Free from the VFS, updating the Usage cache if it is stale;
2. compute newFree = Free ± file.Size;
3. if newFree < 0, abort the operation with an error;
4. if newFree ⩾ 0, let the operation proceed, _and_ update Free and Used in the VFS Usage cache.

The problems I have are:

* Where should I “plug” my code? It seems to me, that `fs/operations.Copy` and `fs/operations.DeleteFileWithBackupDir` are appropriate, although `vfs/write.writeAt` and `vfs/write.close` are interesting as well.
* How should I interact with the VFS Usage cache, which seems to be strictly private at the moment? Is it OK to add a function in `vfs.go` that will force new values for Free and Used?

**Comment by tYYGH** [comment]
I made several attempts, and failed :-(
I’m lost in rclone layers: fs, vfs, file, operations… and in rclone file instances: in-memory copy, in-cache copy, remote copy…

I’m only interested in mounted remotes with `--vfs-cache-mode` ⩾ `write`, which should simplify the problem, except I don’t know what code is called on what situation.

Besides, I’m not fully convinced, that using the VFS Statsfs data as a reference to decide whether or not to allow writes is the right choice, because the VFS layer is not accessible from the operations layer. So my attempts so far were made by patching `vfs/write.go` and `vfs/read_write.go`.

I don’t completely abandon yet, but let’s assume, that I'm back to requesting this feature from people with more knowledge of rclone, and filesystems in general…

**Comment by ivandeex** [comment]
The same was requested by #5276
and https://forum.rclone.org/t/mount-free-space/23742

**Comment by KaspianDev** [comment]
https://forum.rclone.org/t/define-max-storage-for-cloud/30550/2

**Comment by ncw** [comment]
This was implemented in 62bcc84f6fc437e310599d4e5b4692f45bab0d4a

    vfs: add --vfs-disk-space-total-size option to manually set the total disk space
    
    Now you can specify --vfs-disk-space-total-size to set the total disk
    space (default to -1)
